### PR TITLE
feat: Create TOTP node

### PR DESCRIPTION
**Pull request by ivov**
https://linear.app/n8n/issue/ADO-517/totp-node

### CODE REVIEW

File: packages/nodes-base/nodes/Totp/Totp.node.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+	async execute(this: IExecuteFunctions): Promise<INodeExecutionData[][]> {
+		const items = this.getInputData();
+		const returnData: INodeExecutionData[] = [];
+
+		const operation = this.getNodeParameter('operation', 0);
+		const credentials = (await this.getCredentials('totpApi')) as { label: string; secret: string };
+
+		const totp = new OTPAuth.TOTP({
+			issuer: 'GitHub',
```
Comment (netroy) [inline packages/nodes-base/nodes/Totp/Totp.node.ts:56]: should `issuer` be a property on the credential as well?

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -11056,7 +11059,7 @@ packages:
   /eslint-import-resolver-node@0.3.6:
     resolution: {integrity: sha512-0En0w03NRVMn9Uiyn8YRPDKvWjxCWkslUEhGNTdGx15RvPJYQ+lbOlqrlNI2vEAs4pDYK4f/HN2TbDmk5TP0iw==}
     dependencies:
-      debug: 3.2.7(supports-color@8.1.1)
+      debug: 3.2.7(supports-color@5.5.0)
```
Comment (netroy) [inline pnpm-lock.yaml:11062]: these changes seem unrelated to the package you added.

File: packages/nodes-base/nodes/Totp/Totp.node.ts (hunk 1 above)
Comment (ivov) [reply packages/nodes-base/nodes/Totp/Totp.node.ts]: Yes, discussed in channel

File: pnpm-lock.yaml (hunk 2 above)
Comment (ivov) [reply pnpm-lock.yaml]: Thanks! Recreated, not sure where those came from.

**Review by netroy — COMMENTED** [review]
Can you also please add some tests for the node ? 🙏🏽

File: packages/nodes-base/test/nodes/Helpers.ts [hunk 3]
```diff
@@ -30,6 +30,24 @@ import path from 'path';
 import { tmpdir } from 'os';
 import { isEmpty } from 'lodash';
 
+import { FAKE_CREDENTIALS_DATA } from './FakeCredentialsMap';
```
Comment (ivov) [inline packages/nodes-base/test/nodes/Helpers.ts:33]: From [this draft](https://github.com/n8n-io/n8n/pull/5455)

File: packages/nodes-base/nodes/Totp/Totp.node.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+					{
+						name: 'Generate Secret',
+						value: 'generateSecret',
+						action: 'Generate secret',
+					},
+				],
+				default: 'generateSecret',
+			},
+			{
+				displayName: 'Additional Options',
```
Comment (michael-radency) [inline packages/nodes-base/nodes/Totp/Totp.node.ts:48]: Should it be just an 'Options'?

File: packages/nodes-base/nodes/Totp/Totp.node.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+		if (!credentials.label.includes(':')) {
+			throw new NodeOperationError(this.getNode(), 'Malformed label - expected `issuer:username`');
+		}
+
+		const additionalOptions = this.getNodeParameter('additionalOptions', 0) as {
+			algorithm: string;
+			digits: number;
+			period: number;
+		};
```
Comment (michael-radency) [inline packages/nodes-base/nodes/Totp/Totp.node.ts:139]: Is it make sense to do this in loop for each incoming item?

File: packages/nodes-base/nodes/Totp/Totp.node.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+			label: credentials.label,
+			secret: credentials.secret,
+			algorithm: additionalOptions.algorithm,
+			digits: additionalOptions.digits,
+			period: additionalOptions.period,
+		});
+
+		const token = totp.generate();
+
+		const secondsRemaining = (30 * (1 - ((Date.now() / 1000 / 30) % 1))) | 0;
```
Comment (michael-radency) [inline packages/nodes-base/nodes/Totp/Totp.node.ts:154]: not exactly sure what it is going on here, but should options.period was taken in account here?

File: packages/nodes-base/nodes/Totp/Totp.node.ts (hunk 5 above)
Comment (ivov) [reply packages/nodes-base/nodes/Totp/Totp.node.ts:139]: Decided against this because token generation depends also on cred data, which is fixed.

File: packages/nodes-base/nodes/Totp/Totp.node.ts (hunk 4 above)
Comment (ivov) [reply packages/nodes-base/nodes/Totp/Totp.node.ts:48]: Not sure what the convention is? Changed it just in case.

File: packages/nodes-base/nodes/Totp/Totp.node.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+		const options = this.getNodeParameter('options', 0) as {
+			algorithm?: string;
+			digits?: number;
+			period?: number;
+		};
+
+		if (!options.algorithm) options.algorithm = 'SHA1';
+		if (!options.digits) options.digits = 6;
+		if (!options.period) options.period = 30;
```
Comment (ivov) [inline packages/nodes-base/nodes/Totp/Totp.node.ts:143]: Is there a cleaner way to specify default options? Adding `default: { algorithm: 'SHA1', period: 30, digits: 6 }` causes the options to start off revealed, which is not what we want, but otherwise `options` defaults to `{}`, which is also not what we want.

Comment (ivov) [reply packages/nodes-base/nodes/Totp/Totp.node.ts:143]: Sorry I should've been clearer. By "cleaner" I meant to avoid repeating all these defaults that were already specified in the node params above, for maintainability.

Comment (michael-radency) [reply packages/nodes-base/nodes/Totp/Totp.node.ts:143]: not that I'm aware of, we could use some const/enum `defaultOptions` to make it dryer

Comment (ivov) [reply packages/nodes-base/nodes/Totp/Totp.node.ts:143]: Thanks! I think the ideal would be for `getNodeParameter` to return an object with every field set to its default value, so that making it dryer does not introduce too much indirection. Let's ignore for now then.

**Review by michael-radency — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@0.224.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.224.0)
